health monitoring

bomanaps opened the pull request:
## What was wrong?

Issue #

## How was it fixed?
Comprehensive Health Metrics: Implemented ConnectionHealth dataclass with latency tracking, success rates, bandwidth monitoring, and weighted health scoring.

Proactive Monitoring Service: Built ConnectionHealthMonitor with periodic health checks, real-time ping measurements, and automatic unhealthy connection replacement.

Health-Aware Load Balancing: Added four intelligent connection selection strategies, including health_based and latency_based, for optimal traffic routing.

API Consistency Fix: Extended new_host() to accept connection_config, resolving previous API inconsistency while maintaining full backward compatibility.
Summary of approach.

### To-Do

- [x] Clean up commit history  
- [x] Add or update documentation related to these changes  
- [x] Add entry to the [release notes](https://github.com/libp2p/py-libp2p/blob/main/newsfragments/README.md)  

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](<>)

bomanaps commented:
Hello @sumanjeet0012 @lla-dane @sukhman-sukh please can i get a review on this, thanks

bomanaps commented:
> @bomanaps Could you please share the link to the specification file or a reference to this feature in other libp2p implementations? It would be easier for me to understand what exactly we want to implement through this PR.

https://github.com/libp2p/py-libp2p/discussions/717#discussioncomment-14270731

bomanaps commented:
@acul71 i had to create a new pr because there was a lot of changes that affected me when i git pulled

bomanaps commented:
Hello @acul71 @sumanjeet0012 @lla-dane @sukhman-sukh please can i get a review on this, thanks

sumanjeet0012 commented on libp2p/__init__.py:
```diff
@@ -296,12 +297,24 @@ def new_host(
+    # Determine which connection config to use
+    effective_connection_config: ConnectionConfig | QUICTransportConfig | None = None
+    if enable_quic and quic_transport_opt is not None:
+        effective_connection_config = quic_transport_opt
```
when quic is enabled, why we are ignoring the connection config ?
can health monitoring work with quic enabled ?

sumanjeet0012 commented on libp2p/network/swarm.py:
```diff
@@ -760,6 +826,185 @@ async def notify_all(self, notifier: Callable[[INotifee], Awaitable[None]]) -> N
+
+        return "\n".join(metrics)
+
+    async def get_health_monitor_status(self) -> dict[str, Any]:
```
Can we create a dataclass and use it instead of using `Any` ?

sumanjeet0012 reviewed (commented):
@bomanaps Would you be able to share a basic screencast demonstrating the execution of health_monitoring_example.py? I'm currently facing some issues with dependency installation and haven't been able to run the example locally.

acul71 commented:
@bomanaps Are the failing test related or not ? do they fail on your local box with `make pr` ?

bomanaps commented:
@seetadev merged something that caused ci to fail, this checked out https://github.com/libp2p/py-libp2p/pull/915/commits/d2c91f1aee5eaf90fdfc06f6c0a50c606f18c737

bomanaps commented:
> > @seetadev merged something that caused ci to fail, this checked out [d2c91f1](https://github.com/libp2p/py-libp2p/commit/d2c91f1aee5eaf90fdfc06f6c0a50c606f18c737)
> 
> Maybe they randomly failed maybe are related. In my local linux box everything work

If we can re run the ci that will be good the logs iis not from my end make pr passes too, but the ci fail is showing test_provide_and_find_providers is checking that envelope_a

bomanaps commented:
@acul71 @sumanjeet0012

acul71 commented:
@sumanjeet0012 LGTM see if you dig it.

seetadev commented:
@bomanaps : Great, thanks Mercy. This PR is indeed ready for final review + merge.

CCing @acul71 and @sumanjeet0012 for their feedback and pointers. Will merge this soon once they give a thumbs up to this comment.

@bomanaps : In the meantime, please try and complete the other PRs that you are currently contributing before the maintainer's call. We plan to do a py-libp2p release once web sockets PR is complete. Would like your opened PRs to be reviewed, merged and landed in the new release too.

sumanjeet0012 commented:
@seetadev The PR looks good to me,
I have tested it and all tests are passing locally, Don't know why failing in CI.

bomanaps commented:
> @sumanjeet0012 , @yashksaini-coder , @bomanaps and @acul71 : Please refer to #954
> 
> @bomanaps : Please also resolve the merge conflicts.

Please can we take a look at this now

seetadev commented:
@bomanaps : Wonderful, thanks for resolving the merge conflicts. Appreciate it.

I have updated the branch and re-ran the CI/CD pipeline.

We will review this PR in detail. Will discuss with @pacrob and have his review after discussing this PR in tomorrow's maintainer's call. 

In the meantime, wish if you could add a discussion page on this PR covering implementation steps, your approach towards the PR, feedback points from @acul71 and @sumanjeet0012. Wish if you could also record a small screencast too. 

We'll do a final review + merge soon.

pacrob commented on libp2p/network/config.py:
```diff
@@ -34,37 +34,107 @@ class RetryConfig:
+    min_health_threshold: float = 0.3  # 0.0 to 1.0
+    min_connections_per_peer: int = 1
+
+    # Health scoring weights
+    latency_weight: float = 0.4
+    success_rate_weight: float = 0.4
+    stability_weight: float = 0.2
```
We introduce these weights here, but they don't appear to be used anywhere.

seetadev commented:
@bomanaps , @acul71 : Updated the branch and re-ran CI/CD pipeline. Wish if you could reply to @pacrob 's questions and feedback points.

@pacrob : Thank you so much for your review and feedback. Appreciate it.

pacrob commented on libp2p/network/health/monitor.py:
```diff
@@ -0,0 +1,360 @@
+        try:
+            # Skip checks during connection warmup window
+            warmup = getattr(self.config, "health_warmup_window", 0.0)
+            if warmup and hasattr(conn, "established_at"):
```
This line checks if an `INetConn` has an `established_at` attribute, but that attribute is never defined for either `INetConn` or `SwarmConn`, rendering the code in this section effectively dead.

bomanaps commented:
> @pacrob : Thank you so much for sharing the feedback pointers and pointers. Appreciate it.
> 
> @bomanaps — Wish if we could prioritize on the feedback shared by @pacrob:
> 
>     * Apply the style/nit changes (e.g. use @Property for boolean helpers).
> 
>     * More importantly, make sure we add automated tests for the health monitoring logic. We don’t want to merge this much new code without solid regression coverage.
> 
>     * Review all the recommended changes suggested in important sections of the code.
> 
> 
> Once that’s in place, we can revisit additional test case scenarios (like the metrics test coverage checklist) to ensure we’re aligned across the stack.
> 
> @acul71 : The review & detailed notes by @pacrob are very helpful. We should coordinate with @bomanaps together to arrive at a good conclusion on the feedback shared by Paul.

I did address the feedback, it is in this commit https://github.com/libp2p/py-libp2p/pull/915/commits/9c7c61b3bd1b5f7dabcd81b8cf494514f04259d0

yashksaini-coder commented:
> > @sumanjeet0012 , @yashksaini-coder , @bomanaps and @acul71 : Please refer to #954
> > @bomanaps : Please also resolve the merge conflicts.
> 
> Please can we take a look at this now

@bomanaps could you add me to this project please

bomanaps commented:
@acul71 @seetadev

seetadev commented:
@bomanaps : From a release notes perspective, wish if we could change the title of the PR to: Enhance Connection Reliability with Health Metrics, Monitoring in Py-libp2p. Just a suggestion.

pacrob commented on tests/core/transport/test_websocket.py:
```diff
@@ -7,10 +7,10 @@
 else:
     # Fallback for older Python versions
-    ExceptionGroup = Exception
+    ExceptionGroup = Exception  # type: ignore[misc]
```
Removing these type ignores doesn't cause any error for me locally. Are they still necessary?

pacrob commented on libp2p/network/health/monitor.py:
```diff
@@ -0,0 +1,365 @@
+
+        return False
+
+    async def _replace_unhealthy_connection(
```
In this method, we first check if removing the bad connection would drop us below `min_connections_per_peer`, and only replace that bad connection if it won't.

But what if we are sitting at or below that threshold? Then we would just sit with a bad connection and never try to replace it. Would it make sense to try to try to establish a new connection with the peer first, and then drop the bad connection if it's successful?

seetadev commented:
@bomanaps : Wish if you could resolve merge conflicts listed above.

acul71 commented:
@bomanaps I've merged with origin main, see if everything is in order.

This is a detailed review, let's talk about that, and write what you'll address..
Ciao Luca


# AI Pull Request Review: PR #915 - Health Monitoring

**Review Date:** 2025-11-18  
**PR:** [#915](https://github.com/libp2p/py-libp2p/pull/915)  
**Author:** @bomanaps  
**Branch:** `feature/health-monitoring`

---

## 1. Summary of Changes

This PR implements comprehensive connection health monitoring and intelligent load balancing for py-libp2p. The implementation addresses the 1-1 connection mapping limitation discussed in [Discussion #717](https://github.com/libp2p/py-libp2p/discussions/717).

### Issues Addressed
- **Discussion #717**: Connection health monitoring and multi-connection management
- **API Inconsistency**: Extends `new_host()` to accept `connection_config` parameter, resolving previous API inconsistency

### Modules Affected
- **`libp2p/network/health/`**: New module with health monitoring data structures and service
  - `data_structures.py`: ConnectionHealth dataclass with comprehensive metrics tracking
  - `monitor.py`: ConnectionHealthMonitor service for proactive monitoring
- **`libp2p/network/swarm.py`**: Integration of health monitoring into connection lifecycle
- **`libp2p/network/config.py`**: Extended ConnectionConfig with health monitoring options
- **`libp2p/__init__.py`**: API fix to accept `connection_config` in `new_host()`
- **`libp2p/abc.py`**: Added health monitoring interface methods to INetwork and IHost
- **`libp2p/host/basic_host.py`**: Implementation of health monitoring host API methods
- **Documentation**: New example files and ReST documentation

### Breaking Changes
None. The implementation maintains full backward compatibility. Health monitoring is opt-in via configuration.

### Statistics
- **Files Changed:** 29 files
- **Additions:** 4,795 lines
- **Deletions:** 35 lines
- **Test Files:** 4 new test files with 80+ tests

---

## 2. Strengths

### Architecture & Design
- **Well-structured module organization**: Health monitoring is cleanly separated into its own module with clear responsibilities
- **Comprehensive data structures**: `ConnectionHealth` dataclass tracks extensive metrics (latency, success rates, bandwidth, error history, stability)
- **Service-based design**: Uses the existing `Service` base class for proper lifecycle management
- **Configurable health scoring**: Weighted algorithm allows customization of health evaluation criteria

### Code Quality
- **Extensive test coverage**: 4 dedicated test files covering data structures, monitor service, host API, and swarm integration
- **Proper async handling**: Correct use of `trio` for async operations and cancellation handling
- **Error handling**: Comprehensive exception handling with appropriate logging
- **Type hints**: Good type annotation coverage throughout

### API Design
- **Backward compatible**: All changes are opt-in via configuration
- **API consistency fix**: Resolves the missing `connection_config` parameter in `new_host()`
- **Clean integration**: Health monitoring integrates seamlessly with existing connection lifecycle
- **Multiple load balancing strategies**: Supports round_robin, least_loaded, health_based, and latency_based

### Documentation
- **Comprehensive examples**: Two example files demonstrating health monitoring usage
- **ReST documentation**: Proper documentation structure for Sphinx
- **Well-documented code**: Good docstrings throughout the implementation

### Testing
- **Test suite passes**: All 1689 tests pass (4 skipped, 1 warning unrelated to this PR)
- **Comprehensive coverage**: Tests cover normal operations, error cases, edge cases, and integration scenarios
- **Proper async testing**: Uses pytest-asyncio correctly

---

## 3. Issues Found

### Critical

None. The critical issues mentioned in previous reviews have been addressed:
- ✅ Connection replacement logic now uses `dial_peer_replacement()` which creates new connections
- ✅ Warmup logic is functional - `established_at` exists in `ConnectionHealth` dataclass

### Major

#### 3.1 Race Condition in Minimum Connection Handling
- **File:** `libp2p/network/health/monitor.py`
- **Line(s):** 289-295
- **Issue:** When a peer has exactly `min_connections_per_peer` connections and one becomes unhealthy, the code refuses to replace it, potentially leaving an unhealthy connection indefinitely.
- **Suggestion:** Consider allowing replacement even at minimum if the connection is critically unhealthy (e.g., health_score < 0.1 or ping_success_rate < 0.3), then immediately attempt to establish a replacement. Alternatively, log this as a warning that requires manual intervention.

```python
# Current code:
if remaining_after_removal < self.config.min_connections_per_peer:
    logger.warning(...)
    return
```

#### 3.2 Missing @abstractmethod Decorators (Design Question)
- **File:** `libp2p/abc.py`
- **Line(s):** 1636, 1654, 1667, 1685, 2005, 2023, 2036, 2054
- **Issue:** Health monitoring methods in `INetwork` and `IHost` interfaces have default implementations returning empty dicts/strings rather than being marked as `@abstractmethod`. This means implementations are not required to provide these methods.
- **Suggestion:** This is a design decision. If health monitoring is optional, the current approach is acceptable. However, if all implementations should support health monitoring (even if disabled), consider making these abstract. Alternatively, document that these are optional methods with default "disabled" behavior.

#### 3.3 Connection Replacement Not Added to Swarm Connections
- **File:** `libp2p/network/health/monitor.py`
- **Line(s):** 316-324
- **Issue:** After successfully establishing a replacement connection via `dial_peer_replacement()`, the new connection may not be automatically added to the swarm's connection tracking if `dial_peer_replacement()` doesn't handle this internally.
- **Suggestion:** Verify that `dial_peer_replacement()` properly adds the connection to swarm tracking. If not, explicitly add it after successful establishment and initialize health tracking for it.

### Minor

#### 3.4 Inconsistent Error Handling in Ping
- **File:** `libp2p/network/health/monitor.py`
- **Line(s):** 195-225
- **Issue:** The `_ping_connection()` method returns `True` if there are active streams without actually pinging, which may mask connection issues.
- **Suggestion:** Consider still performing a lightweight ping even when streams are active, or document this behavior clearly.

#### 3.5 Potential Memory Leak in Error History
- **File:** `libp2p/network/health/data_structures.py`
- **Line(s):** 163-170
- **Issue:** Error history is capped at 100 entries, but if errors occur frequently, this could still grow large over time for long-lived connections.
- **Suggestion:** Consider time-based cleanup (e.g., only keep errors from last 24 hours) in addition to count-based cleanup.

#### 3.6 Missing Validation in Health Score Calculation
- **File:** `libp2p/network/health/data_structures.py`
- **Line(s):** 110-121
- **Issue:** The `update_health_score()` method normalizes latency to 1 second, but doesn't handle edge cases where latency might be negative or extremely large.
- **Suggestion:** Add bounds checking and handle edge cases (e.g., if latency > 1000ms, set latency_score to 0.0).

---

## 4. Security Review

### Security Assessment: **Low Risk**

No significant security vulnerabilities identified. The implementation follows secure coding practices:

- ✅ **No external input validation issues**: All configuration comes from trusted sources (application code)
- ✅ **No unsafe subprocess usage**: No subprocess calls present
- ✅ **Proper key handling**: No cryptographic operations in this module
- ✅ **No sensitive data exposure**: Health metrics don't expose sensitive information
- ✅ **Safe defaults**: Health monitoring is opt-in, disabled by default

### Minor Security Considerations

- **Information Disclosure**: Health metrics could potentially reveal network topology or connection patterns. This is expected behavior for monitoring but should be documented.
- **Resource Exhaustion**: Health monitoring adds overhead. The implementation includes reasonable limits (error history capped, bandwidth windows limited), but should be monitored in production.

---

## 5. Documentation and Examples

### Documentation Quality: **Excellent**

- ✅ **Newsfragment present**: `newsfragments/915.feature.rst` exists and is well-formatted
- ✅ **Example files**: Two comprehensive examples provided:
  - `examples/health_monitoring_example.py`: Basic usage
  - `examples/health_monitoring_quic_example.py`: QUIC-specific usage
- ✅ **ReST documentation**: `docs/examples.connection_health_monitoring.rst` provides structured documentation
- ✅ **Code docstrings**: Comprehensive docstrings throughout
- ✅ **API documentation**: Methods are properly documented in `abc.py`

### Documentation Suggestions

1. **Usage Guide**: Consider adding a "Getting Started with Health Monitoring" section to the main documentation
2. **Configuration Reference**: Document all `ConnectionConfig` parameters in a centralized configuration reference
3. **Troubleshooting**: Add a section on common issues (e.g., connections not being replaced, high false positive rates)

---

## 6. Newsfragment Requirement

### Status: ✅ **PASS**

- **File:** `newsfragments/915.feature.rst`
- **Format:** Correct (`915.feature.rst`)
- **Content:** Comprehensive user-facing description
- **Newline:** Present (verified in file)
- **Type:** Appropriate (`feature` for new functionality)

The newsfragment is well-written and describes the feature from a user perspective, covering:
- Connection health metrics
- Proactive monitoring service
- Health-aware load balancing
- API consistency fix
- Configuration and integration

---

## 7. Tests and Validation

### Test Results: ✅ **ALL PASSING**

- **Linting:** ✅ Passed (ruff, black, isort, pyupgrade, mdformat)
- **Type Checking:** ✅ Passed (mypy, pyrefly)
- **Test Execution:** ✅ **1689 passed, 4 skipped, 1 warning** (unrelated to this PR)
- **Documentation Build:** ✅ Passed (Sphinx build successful)

### Test Coverage

**New Test Files:**
1. `tests/core/network/test_health_data_structures.py` - Tests for ConnectionHealth dataclass
2. `tests/core/network/test_health_monitor.py` - Tests for ConnectionHealthMonitor service
3. `tests/core/network/test_health_host_api.py` - Tests for host API integration
4. `tests/core/network/test_health_swarm_integration.py` - Integration tests with Swarm

**Test Quality:**
- ✅ Covers normal operations
- ✅ Covers error cases
- ✅ Covers edge cases (minimum connections, warmup windows, etc.)
- ✅ Proper async test patterns
- ✅ Good use of fixtures and mocks

### Test Recommendations

1. **Race Condition Tests**: Add specific tests for the minimum connection edge case
2. **Concurrent Replacement Tests**: Test behavior when multiple connections to the same peer become unhealthy simultaneously
3. **Load Balancing Strategy Tests**: Verify all four load balancing strategies work correctly under various scenarios

---

## 8. Recommendations for Improvement

### High Priority

1. **Address Race Condition (Issue 3.1)**: Implement logic to allow replacement at minimum connections for critically unhealthy connections
2. **Verify Connection Tracking (Issue 3.3)**: Ensure replacement connections are properly tracked in swarm

### Medium Priority

3. **Document Design Decision (Issue 3.2)**: Clarify whether health monitoring methods should be abstract or optional with defaults
4. **Enhance Ping Logic (Issue 3.4)**: Consider pinging even when streams are active, or document the current behavior
5. **Add Time-Based Cleanup (Issue 3.5)**: Implement time-based error history cleanup

### Low Priority

6. **Improve Edge Case Handling (Issue 3.6)**: Add bounds checking in health score calculation
7. **Add Performance Metrics**: Consider adding metrics for health monitoring overhead
8. **Configuration Validation**: Add runtime validation warnings for potentially problematic configurations

---

## 9. Questions for the Author

1. **Connection Replacement at Minimum**: What is the intended behavior when a peer has exactly `min_connections_per_peer` connections and one becomes unhealthy? Should we always maintain the minimum, or allow replacement for critically unhealthy connections?

2. **Abstract Methods**: Was the decision to use default implementations (rather than `@abstractmethod`) for health monitoring methods intentional? Should all implementations be required to support health monitoring, or is it truly optional?

3. **Ping Strategy**: Why does `_ping_connection()` skip pinging when streams are active? Is this a performance optimization, or is there another reason?

4. **Load Balancing**: Are there plans to add more sophisticated load balancing strategies (e.g., weighted round-robin based on health scores)?

5. **Metrics Export**: The `export_health_metrics()` method supports "json" and "prometheus" formats, but only JSON is implemented. Are Prometheus exports planned for a future PR?

---

## 10. Overall Assessment

### Quality Rating: **Good**

The implementation is well-structured, comprehensive, and addresses the core requirements. The code quality is high with good test coverage and documentation. Most critical issues from previous reviews have been addressed.

### Security Impact: **None**

No security vulnerabilities identified. The implementation follows secure coding practices.

### Merge Readiness: **Needs Minor Fixes**

The PR is close to merge-ready but should address:
1. The race condition in minimum connection handling (Issue 3.1)
2. Verification that replacement connections are properly tracked (Issue 3.3)

These are relatively minor issues that can be addressed quickly.

### Confidence: **High**

The implementation is solid and well-tested. The remaining issues are edge cases and design clarifications rather than fundamental problems.

---

## Additional Notes

### Previous Review Feedback Status

From @pacrob's review, the following issues have been **RESOLVED**:
- ✅ Connection replacement logic bug - Fixed with `dial_peer_replacement()`
- ✅ Dead code in warmup logic - Not dead code, `established_at` exists and is used correctly

The following issues remain as **DESIGN DECISIONS** or **MINOR IMPROVEMENTS**:
- ⚠️ Race condition handling - Addressed in Issue 3.1 above
- ⚠️ Missing @abstractmethod decorators - Addressed in Issue 3.2 above (may be intentional design)

### Commendations

- Excellent test coverage with 80+ new tests
- Comprehensive documentation and examples
- Clean module organization
- Proper async/await patterns throughout
- Good error handling and logging

---

acul71 commented:
> > @bomanaps : Hi Mercy. Appreciate your continued momentum. Wish if we could resolve the merge conflicts and CI/CD issues.
> > Also, re-adding @acul71, @yashksaini-coder to the thread along with @sumanjeet0012, who will help you arrive at a good conclusion on the PR with their peer reviews and feedback.
> 
> I have addressed the conflicts

Hi @bomanaps here's the review. 
Open an issue and update the PR desc and newsfragment number.
Then take a look at others issues here: 


## 1. Summary of Changes

This PR implements comprehensive connection health monitoring and intelligent load balancing for py-libp2p. The implementation addresses the 1-1 connection mapping limitation discussed in [Discussion #717](https://github.com/libp2p/py-libp2p/discussions/717).

### Issues Addressed
- **Discussion #717**: Connection health monitoring and multi-connection management (addresses the TODO about 1-1 connection mapping limitation)
- **API Inconsistency**: Extends `new_host()` to accept `connection_config` parameter, resolving previous API inconsistency

⚠️ **CRITICAL BLOCKER**: The PR body shows "Issue #" without a number. According to project requirements, **every PR MUST refer to a GitHub issue**. This is a mandatory requirement. The PR author must open an issue first before the PR can be approved.

### Modules Affected
- **`libp2p/network/health/`**: New module with health monitoring data structures and service
  - `data_structures.py`: ConnectionHealth dataclass with comprehensive metrics tracking
  - `monitor.py`: ConnectionHealthMonitor service for proactive monitoring
- **`libp2p/network/swarm.py`**: Integration of health monitoring into connection lifecycle
- **`libp2p/network/config.py`**: Extended ConnectionConfig with health monitoring options
- **`libp2p/__init__.py`**: API fix to accept `connection_config` in `new_host()`
- **`libp2p/abc.py`**: Added health monitoring interface methods to INetwork and IHost
- **`libp2p/host/basic_host.py`**: Implementation of health monitoring host API methods
- **Documentation**: New example files and ReST documentation

### Breaking Changes
None. The implementation maintains full backward compatibility. Health monitoring is opt-in via configuration.

### Statistics
- **Files Changed:** 29 files
- **Additions:** 4,483 lines
- **Deletions:** 26 lines
- **Test Files:** 4 new test files with 80+ tests
- **Commits:** 21 commits

---

## 2. Branch Sync Status and Merge Conflicts

### Branch Sync Status
- **Status:** ✅ **In sync** (`mergeStateStatus: "CLEAN"`, `mergeable: "MERGEABLE"`)
- **Details:** PR has 21 commits and is in sync with main (`mergeStateStatus: "CLEAN"`). The PR branch is ready to merge.

### Merge Conflict Analysis
- **Conflicts Detected:** ✅ **No conflicts** - PR can be merged cleanly
- **Test Merge Result:** The test merge completed successfully with "Already up to date" message, confirming no conflicts exist.

---

## 3. Strengths

### Architecture & Design
- **Well-structured module organization**: Health monitoring is cleanly separated into its own module (`libp2p/network/health/`) with clear responsibilities
- **Comprehensive data structures**: `ConnectionHealth` dataclass tracks extensive metrics (latency, success rates, bandwidth, error history, stability)
- **Service-based design**: Uses the existing `Service` base class for proper lifecycle management
- **Configurable health scoring**: Weighted algorithm allows customization of health evaluation criteria (latency_weight, success_rate_weight, stability_weight)
- **Multiple load balancing strategies**: Supports round_robin, least_loaded, health_based, and latency_based strategies

### Code Quality
- **Extensive test coverage**: 4 dedicated test files covering data structures, monitor service, host API, and swarm integration (80+ new tests)
- **Proper async handling**: Correct use of `trio` for async operations and cancellation handling
- **Error handling**: Comprehensive exception handling with appropriate logging
- **Type hints**: Good type annotation coverage throughout
- **Clean integration**: Health monitoring integrates seamlessly with existing connection lifecycle

### API Design
- **Backward compatible**: All changes are opt-in via configuration
- **API consistency fix**: Resolves the missing `connection_config` parameter in `new_host()`
- **Clean integration**: Health monitoring integrates seamlessly with existing connection lifecycle
- **Host-level API**: Health monitoring accessible through high-level host interface methods

### Documentation
- **Comprehensive examples**: Two example files demonstrating health monitoring usage
- **ReST documentation**: Proper documentation structure for Sphinx (`docs/examples.connection_health_monitoring.rst`)
- **Well-documented code**: Good docstrings throughout the implementation
- **Newsfragment present**: `newsfragments/915.feature.rst` exists and is well-formatted

### Testing
- **Test suite passes**: All 1945 tests pass (15 skipped, 25 warnings unrelated to this PR)
- **Comprehensive coverage**: Tests cover normal operations, error cases, edge cases, and integration scenarios
- **Proper async testing**: Uses pytest-asyncio correctly

---

## 4. Issues Found

### Critical

#### 4.1 Missing Issue Reference (BLOCKER)
- **File:** PR description
- **Issue:** The PR body shows "Issue #" without a number. According to project requirements, **every PR MUST refer to a GitHub issue**. This is a mandatory requirement, not optional.
- **Impact:** 
  - PR cannot be approved without a linked issue
  - Newsfragment is incorrectly named using PR number (`915.feature.rst`) instead of issue number
- **Suggestion:** 
  - The PR author must open a GitHub issue first to describe the feature, bug fix, or change
  - Once an issue exists, update the PR description to reference it (e.g., "Fixes #XXX", "Closes #XXX", "Implements #XXX")
  - **Rename the newsfragment** from `newsfragments/915.feature.rst` to `newsfragments/<ISSUE_NUMBER>.feature.rst` where `<ISSUE_NUMBER>` is the actual issue number
  - Note: Discussion #717 is referenced, but a GitHub issue is still required per project policy
- **Action Required:** PR approval is blocked until:
  1. An issue is opened and linked
  2. The newsfragment is renamed to use the issue number

### Major

#### 4.1 Race Condition in Minimum Connection Handling
- **File:** `libp2p/network/health/monitor.py`
- **Line(s):** 289-295 (approximate, based on previous reviews)
- **Issue:** When a peer has exactly `min_connections_per_peer` connections and one becomes unhealthy, the code refuses to replace it, potentially leaving an unhealthy connection indefinitely.
- **Suggestion:** Consider allowing replacement even at minimum if the connection is critically unhealthy (e.g., health_score < 0.1 or ping_success_rate < 0.3), then immediately attempt to establish a replacement. Alternatively, log this as a warning that requires manual intervention.

#### 4.2 Missing @abstractmethod Decorators (Design Question)
- **File:** `libp2p/abc.py`
- **Line(s):** Health monitoring methods in `INetwork` and `IHost` interfaces
- **Issue:** Health monitoring methods in `INetwork` and `IHost` interfaces have default implementations returning empty dicts/strings rather than being marked as `@abstractmethod`. This means implementations are not required to provide these methods.
- **Suggestion:** This is a design decision. If health monitoring is optional, the current approach is acceptable. However, if all implementations should support health monitoring (even if disabled), consider making these abstract. Alternatively, document that these are optional methods with default "disabled" behavior.

#### 4.3 Connection Replacement Verification
- **File:** `libp2p/network/health/monitor.py`
- **Line(s):** Connection replacement logic
- **Issue:** After successfully establishing a replacement connection via `dial_peer_replacement()`, verify that the new connection is properly added to swarm tracking and health tracking is initialized.
- **Suggestion:** Verify that `dial_peer_replacement()` properly adds the connection to swarm tracking. If not, explicitly add it after successful establishment and initialize health tracking for it.

### Minor

#### 4.4 Inconsistent Error Handling in Ping
- **File:** `libp2p/network/health/monitor.py`
- **Line(s):** `_ping_connection()` method
- **Issue:** The `_ping_connection()` method returns `True` if there are active streams without actually pinging, which may mask connection issues.
- **Suggestion:** Consider still performing a lightweight ping even when streams are active, or document this behavior clearly.

#### 4.5 Potential Memory Leak in Error History
- **File:** `libp2p/network/health/data_structures.py`
- **Line(s):** Error history management
- **Issue:** Error history is capped at 100 entries, but if errors occur frequently, this could still grow large over time for long-lived connections.
- **Suggestion:** Consider time-based cleanup (e.g., only keep errors from last 24 hours) in addition to count-based cleanup.

#### 4.6 Missing Validation in Health Score Calculation
- **File:** `libp2p/network/health/data_structures.py`
- **Line(s):** `update_health_score()` method
- **Issue:** The `update_health_score()` method normalizes latency to 1 second, but doesn't handle edge cases where latency might be negative or extremely large.
- **Suggestion:** Add bounds checking and handle edge cases (e.g., if latency > 1000ms, set latency_score to 0.0; if latency < 0, set to 0.0).

---

## 5. Security Review

### Security Assessment: **Low Risk**

No significant security vulnerabilities identified. The implementation follows secure coding practices:

- ✅ **No external input validation issues**: All configuration comes from trusted sources (application code)
- ✅ **No unsafe subprocess usage**: No subprocess calls present
- ✅ **Proper key handling**: No cryptographic operations in this module
- ✅ **No sensitive data exposure**: Health metrics don't expose sensitive information
- ✅ **Safe defaults**: Health monitoring is opt-in, disabled by default

### Minor Security Considerations

- **Information Disclosure**: Health metrics could potentially reveal network topology or connection patterns. This is expected behavior for monitoring but should be documented.
- **Resource Exhaustion**: Health monitoring adds overhead. The implementation includes reasonable limits (error history capped, bandwidth windows limited), but should be monitored in production.

---

## 6. Documentation and Examples

### Documentation Quality: **Excellent**

- ✅ **Newsfragment present**: `newsfragments/915.feature.rst` exists and is well-formatted
- ✅ **Example files**: Two comprehensive examples provided:
  - `examples/health_monitoring_example.py`: Basic usage
  - `examples/health_monitoring_quic_example.py`: QUIC-specific usage
- ✅ **ReST documentation**: `docs/examples.connection_health_monitoring.rst` provides structured documentation
- ✅ **Code docstrings**: Comprehensive docstrings throughout
- ✅ **API documentation**: Methods are properly documented in `abc.py`

### Documentation Suggestions

1. **Usage Guide**: Consider adding a "Getting Started with Health Monitoring" section to the main documentation
2. **Configuration Reference**: Document all `ConnectionConfig` parameters in a centralized configuration reference
3. **Troubleshooting**: Add a section on common issues (e.g., connections not being replaced, high false positive rates)

---

## 7. Newsfragment Requirement

### Status: ❌ **FAIL - INCORRECT NAMING**

- **File:** `newsfragments/915.feature.rst`
- **Format:** ❌ **INCORRECT** - Uses PR number (`915.feature.rst`) instead of issue number
- **Content:** ✅ Comprehensive user-facing description
- **Newline:** ✅ Present (verified in file)
- **Type:** ✅ Appropriate (`feature` for new functionality)

**Critical Issue:** The newsfragment file is named using the PR number (915) instead of an issue number. According to project requirements, newsfragments must be named as `<ISSUE_NUMBER>.<TYPE>.rst` (e.g., `910.feature.rst` for issue #910).

**Current State:**
- Newsfragment exists: `newsfragments/915.feature.rst` (uses PR number ❌)
- Should be: `newsfragments/<ISSUE_NUMBER>.feature.rst` (uses issue number ✅)

**Root Cause:** The PR does not reference a GitHub issue, so the newsfragment was incorrectly created using the PR number.

**Required Actions:**
1. **Open a GitHub issue** describing the health monitoring feature (if not already done)
2. **Rename the newsfragment** from `915.feature.rst` to `<ISSUE_NUMBER>.feature.rst` where `<ISSUE_NUMBER>` is the actual issue number
3. **Update the PR description** to reference the issue (e.g., "Fixes #XXX" or "Implements #XXX")

The newsfragment content is well-written and describes the feature from a user perspective, covering:
- Connection health metrics
- Proactive monitoring service
- Health-aware load balancing
- API consistency fix
- Configuration and integration

However, the incorrect naming is a **BLOCKER** because it violates the project's newsfragment naming convention.

---

## 8. Tests and Validation

### Test Results: ✅ **ALL PASSING**

#### Linting (`make lint`)
- **Status:** ✅ **Passed**
- **Exit Code:** 0
- **Details:** All linting checks passed:
  - check yaml: Passed
  - check toml: Passed
  - fix end of files: Passed
  - trim trailing whitespace: Passed
  - pyupgrade: Passed
  - ruff (legacy alias): Passed
  - ruff format: Passed
  - mdformat: Passed
  - run mypy with all dev dependencies present: Passed
  - run pyrefly typecheck locally: Passed
  - Check for .rst files in the top-level directory: Passed
- **No errors, warnings, or issues detected**

#### Type Checking (`make typecheck`)
- **Status:** ✅ **Passed**
- **Exit Code:** 0
- **Details:** All type checking passed:
  - run mypy with all dev dependencies present: Passed
  - run pyrefly typecheck locally: Passed
- **No type errors, warnings, or issues detected**

#### Test Execution (`make test`)
- **Status:** ✅ **PASSED**
- **Exit Code:** 0
- **Total Tests:** 1945
- **Passed:** 1945
- **Failed:** 0
- **Skipped:** 15
- **Warnings:** 25 (unrelated to this PR)
- **Execution Time:** 98.16s (1:38)
- **Details:**
  - All tests pass successfully
  - 15 skipped tests (expected, platform-specific or conditional tests)
  - 25 warnings are unrelated to this PR (mostly pytest.mark.integration warnings and one RuntimeWarning about coroutine)
  - No test failures or errors related to health monitoring changes

#### Documentation Build (`make linux-docs`)
- **Status:** ✅ **Passed**
- **Exit Code:** 0
- **Details:** 
  - Sphinx build successful
  - All 83 source files processed
  - HTML documentation generated successfully
  - Doctest summary: 109 tests, 0 failures
  - Newsfragment validation passed
- **No errors or warnings detected**

### Test Coverage

**New Test Files:**
1. `tests/core/network/test_health_data_structures.py` - Tests for ConnectionHealth dataclass
2. `tests/core/network/test_health_monitor.py` - Tests for ConnectionHealthMonitor service
3. `tests/core/network/test_health_host_api.py` - Tests for host API integration
4. `tests/core/network/test_health_swarm_integration.py` - Integration tests with Swarm

**Test Quality:**
- ✅ Covers normal operations
- ✅ Covers error cases
- ✅ Covers edge cases (minimum connections, warmup windows, etc.)
- ✅ Proper async test patterns
- ✅ Good use of fixtures and mocks

### Test Recommendations

1. **Race Condition Tests**: Add specific tests for the minimum connection edge case
2. **Concurrent Replacement Tests**: Test behavior when multiple connections to the same peer become unhealthy simultaneously
3. **Load Balancing Strategy Tests**: Verify all four load balancing strategies work correctly under various scenarios

---

## 9. Recommendations for Improvement

### High Priority

1. **Address Missing Issue Reference (Issue 4.1)**: This is a CRITICAL BLOCKER. The PR author must:
   - Open a GitHub issue describing the health monitoring feature
   - Update the PR description to reference the issue (e.g., "Fixes #XXX" or "Implements #XXX")
   - **Rename the newsfragment** from `newsfragments/915.feature.rst` to `newsfragments/<ISSUE_NUMBER>.feature.rst` where `<ISSUE_NUMBER>` is the actual issue number
   - This is mandatory per project requirements

2. **Address Race Condition (Issue 4.1)**: Implement logic to allow replacement at minimum connections for critically unhealthy connections

3. **Verify Connection Tracking (Issue 4.3)**: Ensure replacement connections are properly tracked in swarm

### Medium Priority

4. **Document Design Decision (Issue 4.2)**: Clarify whether health monitoring methods should be abstract or optional with defaults

5. **Enhance Ping Logic (Issue 4.4)**: Consider pinging even when streams are active, or document the current behavior

6. **Add Time-Based Cleanup (Issue 4.5)**: Implement time-based error history cleanup

### Low Priority

7. **Improve Edge Case Handling (Issue 4.6)**: Add bounds checking in health score calculation

8. **Add Performance Metrics**: Consider adding metrics for health monitoring overhead

9. **Configuration Validation**: Add runtime validation warnings for potentially problematic configurations

---

## 10. Questions for the Author

1. **Issue Reference**: Can you open a GitHub issue for this PR and update the PR description to reference it? This is a mandatory requirement per project policy. Once the issue is created, the newsfragment file `newsfragments/915.feature.rst` must be renamed to `newsfragments/<ISSUE_NUMBER>.feature.rst` to follow the project's naming convention.

2. **Connection Replacement at Minimum**: What is the intended behavior when a peer has exactly `min_connections_per_peer` connections and one becomes unhealthy? Should we always maintain the minimum, or allow replacement for critically unhealthy connections?

3. **Abstract Methods**: Was the decision to use default implementations (rather than `@abstractmethod`) for health monitoring methods intentional? Should all implementations be required to support health monitoring, or is it truly optional?

4. **Ping Strategy**: Why does `_ping_connection()` skip pinging when streams are active? Is this a performance optimization, or is there another reason?

5. **Load Balancing**: Are there plans to add more sophisticated load balancing strategies (e.g., weighted round-robin based on health scores)?

6. **Metrics Export**: The `export_health_metrics()` method supports "json" and "prometheus" formats, but only JSON is implemented. Are Prometheus exports planned for a future PR?

---

## 11. Overall Assessment

### Quality Rating: **Good**

The implementation is well-structured, comprehensive, and addresses the core requirements. The code quality is high with good test coverage and documentation. Most critical issues from previous reviews have been addressed.

### Security Impact: **None**

No security vulnerabilities identified. The implementation follows secure coding practices.

### Merge Readiness: **Needs Critical Fix**

The PR is close to merge-ready but has a **CRITICAL BLOCKER**:

1. **Missing Issue Reference (BLOCKER)**: The PR must reference a GitHub issue. This is a mandatory requirement per project policy. The PR author must:
   - Open a GitHub issue describing the health monitoring feature
   - Update the PR description to reference the issue
   - **Rename the newsfragment** from `newsfragments/915.feature.rst` to `newsfragments/<ISSUE_NUMBER>.feature.rst` where `<ISSUE_NUMBER>` is the actual issue number

Once the issue reference is added and the newsfragment is renamed, the PR should address:
2. The race condition in minimum connection handling (Issue 4.1)
3. Verification that replacement connections are properly tracked (Issue 4.3)

These are relatively minor issues that can be addressed quickly.

### Confidence: **High**

The implementation is solid and well-tested. The remaining issues are edge cases and design clarifications rather than fundamental problems. The critical blocker (missing issue reference) is a process requirement that can be easily resolved.

---

## Additional Notes

### Previous Review Feedback Status

From previous reviews, the following issues have been **RESOLVED**:
- ✅ Connection replacement logic bug - Fixed with `dial_peer_replacement()`
- ✅ Dead code in warmup logic - Not dead code, `established_at` exists and is used correctly
- ✅ Added 80+ tests - Comprehensive test coverage added
- ✅ Fixed CI failures - All tests passing
- ⚠️ Newsfragment present but incorrectly named - Uses PR number (915) instead of issue number
- ✅ Resolved merge conflicts - No conflicts detected

The following issues remain as **DESIGN DECISIONS** or **MINOR IMPROVEMENTS**:
- ⚠️ Race condition handling - Addressed in Issue 4.1 above
- ⚠️ Missing @abstractmethod decorators - Addressed in Issue 4.2 above (may be intentional design)
- ⚠️ Missing issue reference - **CRITICAL BLOCKER** (Issue 4.1)
- ⚠️ Newsfragment incorrectly named - Uses PR number instead of issue number (Issue 4.1)

### Commendations

- Excellent test coverage with 80+ new tests
- Comprehensive documentation and examples
- Clean module organization
- Proper async/await patterns throughout
- Good error handling and logging
- All linting, type checking, tests, and documentation builds pass

### Discussion Reference

The PR references [Discussion #717](https://github.com/libp2p/py-libp2p/discussions/717), which provides excellent context about the motivation for this feature. The discussion explains the 1-1 connection mapping limitation and the need for health monitoring. However, per project requirements, a GitHub issue is still required in addition to the discussion.

bomanaps commented:
> Hi @bomanaps here's the review. Open an issue and update the PR desc and newsfragment number. Then take a look at others issues here:
> ## 1. Summary of Changes
> 
> This PR implements comprehensive connection health monitoring and intelligent load balancing for py-libp2p. The implementation addresses the 1-1 connection mapping limitation discussed in [Discussion #717](https://github.com/libp2p/py-libp2p/discussions/717).
> ### Issues Addressed
> 
>     * **Discussion **

Hello @acul71 I have addressed the comment you gave on this please review again

pacrob commented on docs/examples.connection_health_monitoring.rst:
```diff
@@ -0,0 +1,292 @@
+        load_balancing_strategy="health_based"  # Use health-based selection
+    )
+
+    # Create host with health monitoring - API consistency fixed!
```
The `API concistency fixed!` looks like an artifact from an update maybe?

pacrob commented on libp2p/__init__.py:
```diff
@@ -411,11 +419,61 @@ def new_host(
+                "max_ping_latency",
+                "min_ping_success_rate",
+                "max_failed_streams",
+                "unhealthy_grace_period",
```
This is missing `critical_health_threshold` from `ConnectionConfig`.

pacrob commented on libp2p/network/config.py:
```diff
@@ -34,37 +34,110 @@ class RetryConfig:
+            if not 0.0 <= self.min_ping_success_rate <= 1.0:
+                raise ValueError("Min ping success rate must be between 0.0 and 1.0")
+            if self.max_failed_streams < 0:
+                raise ValueError("Max failed streams must be non-negative")
```
Missing `critical_health_threshold` validation here.

pacrob commented on docs/examples.connection_health_monitoring.rst:
```diff
@@ -0,0 +1,292 @@
+        # Health monitoring and load balancing happen automatically
+        stream = await host.new_stream(peer_id, ["/echo/1.0.0"])
+
+Example: Advanced Health Monitoring
```
You marked this as resolved without making any changes.
